src/workerd/io/hibernation-manager.c++

```diff
@@ -0,0 +1,216 @@
+  // Note that the HibernatableWebSocket destructor handles removing any references to itself in
+  // `tagToWs`, and even removes the hashmap entry if there are no more entries in the bucket.
+  allWs.clear();
+  KJ_ASSERT(tagToWs.size() == 0, "tagToWs hashmap wasn't cleared.");
+  KJ_ASSERT(allWs.size() == 0 && allWs.empty(), "allWs collection wasn't cleared.");
+}
+
+void HibernationManagerImpl::acceptWebSocket(
+    jsg::Ref<api::WebSocket> ws,
+    kj::ArrayPtr<kj::String> tags) {
+  // First, we create the HibernatableWebSocket and add it to the collection where it'll stay
+  // until it's destroyed.
+  auto hib = kj::heap<HibernatableWebSocket>(kj::mv(ws), tags, countOfAcceptedWebSockets++, *this);
+  HibernatableWebSocket& refToHibernatable = *hib.get();
+  allWs.push_front(kj::mv(hib));
+  refToHibernatable.node = allWs.begin();
+
+  // If the `tags` array is empty (i.e. user did not provide a tag), we skip the population of the
+  // `tagToWs` HashMap below and go straight to initiating the readLoop.
+
+  // It is the caller's responsibility to ensure all elements of `tags` are unique.
+  // TODO(cleanup): Maybe we could enforce uniqueness by using an immutable type that
+  // can only be constructed if the elements in the collection are distinct, ex. "DistinctArray".
+  //
+  // We need to add the HibernatableWebSocket to each bucket in `tagToWs` corresponding to its tags.
+  //  1. Create the entry if it doesn't exist
+  //  2. Fill the TagListItem in the HibernatableWebSocket's tagItems array
+  size_t position = 0;
+  for (auto tag = tags.begin(); tag < tags.end(); tag++, position++) {
+    auto& tagCollection = tagToWs.findOrCreate(*tag, [&tag]() {
+      auto item = kj::heap<TagCollection>(
+          kj::mv(*tag), kj::heap<kj::List<TagListItem, &TagListItem::link>>());
+      return decltype(tagToWs)::Entry {
+          item->tag,
+          kj::mv(item)
+      };
+    });
+    // This TagListItem sits in the HibernatableWebSocket's tagItems array.
+    auto& tagListItem = refToHibernatable.tagItems[position];
+    tagListItem.hibWS = refToHibernatable;
+    tagListItem.tag = tagCollection->tag.asPtr();
+
+    auto& list = tagCollection->list;
+    list->add(tagListItem);
+    // We also give the TagListItem a reference to the list it was added to so the
+    // HibernatableWebSocket can quickly remove itself from the list without doing a lookup
+    // in `tagToWs`.
+    tagListItem.list = *list.get();
+  }
+
+  // Finally, we initiate the readloop for this HibernatableWebSocket.
+  kj::Promise<kj::Maybe<kj::Exception>> readLoopPromise = kj::evalNow([&] {
+    return readLoop(refToHibernatable);
+  }).then([]() -> kj::Maybe<kj::Exception> { return nullptr; },
+          [](kj::Exception&& e) -> kj::Maybe<kj::Exception> { return kj::mv(e); });
+
+  // Give the task to the HibernationManager so it lives long.
+  readLoopTasks.add(readLoopPromise.then(
+      [&refToHibernatable, this](kj::Maybe<kj::Exception>&& e) {
```

Comment:
Mostly a note for myself:

Last time I really looked at this we had assumed we could get an `api::WebSocket` while in this continuation. Now it looks like we should assume we can only access it during the events execution (or _maybe_ after it completes), so a few things need to change.

1. (Unrelated to updates but) we don't want 2 lambdas, not sure why I did that... one is enough.
2. Like in `web-socket.c++`, we need to check if we disconnected/canceled or if something else happened, so we either dispatch a close or an error event (not _just_ an error event).
3. We still want to release the `Native` websocket, but since we don't have access to it we need public methods to do so. We also need to do that _after_ the event is done, so maybe a continuation on the event, or modifying the event itself in `global-scope.c++`.